addon/utils/resolve.js

```diff
@@ -37,7 +37,28 @@ export function resolveReferencesWithRecords(store, references) {
   if (CUSTOM_MODEL_CLASS) {
     return references.map((reference) => {
       if (reference.type) {
-        return store.peekRecord(dasherize(reference.type), reference.id);
+        let normalizedType = dasherize(reference.type);
+        let record = store.peekRecord(normalizedType, reference.id);
+        if (record) {
+          return record;
+        }
+        // If we have a cached recordData with the same id, but we have not seen a record with the same { type, id } pair
+        // We could be a projection, in which case we want to push in a projected record with the new type
+        let cachedRD = store._globalM3RecordDataCache[reference.id];
+        if (!record && cachedRD) {
```

Comment:
Do we still need a check for `record` here?

Reply:
good catch

Reply:
Fixed